### PR TITLE
[BEAM-2314] Add ValidatesRunner test for merging custom windows

### PR DESCRIPTION
**Pull request by echauchot**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`.
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
@aljoscha I reopened a new PR to test that jenkins integration is working again as requested
R: @aljoscha
CC: @jbonofre

### CODE REVIEW

**Comment by aljoscha** [comment]
Pre-commit seems to be failing.

**Comment by aljoscha** [comment]
Run Flink ValidatesRunner

**Comment by aljoscha** [comment]
Run Spark ValidatesRunner

**Comment by aljoscha** [comment]
Run Apex ValidatesRunner

**Comment by aljoscha** [comment]
Run Gearpump ValidatesRunner

**Comment by aljoscha** [comment]
retest this please

**Comment by aljoscha** [comment]
@echauchot I'm afraid you have to introduce an exclusion (similar to `UsesTestStream` or `UsesMapState`) because at least the Spark runner doesn't handle custom merging windows well. Did you see that problem also in the NEXMark tests?

**Comment by echauchot** [comment]
@aljoscha: strange, the pre-commit tests were passing and I changed nothing, I'll dig into it. For your question about Spark runner: it runs NexMark query9 (custom window merging) correctly in both batch and streaming modes. Maybe there is something different from Nexmark query in my `ValidatesRunner` test. Once again, I'll take a look and ping you when it's ready

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by aljoscha** [comment]
The spark tests still seem to be failing. Any news?

Maybe @aviemzur has an idea what could be going on?

**Comment by echauchot** [comment]
Hi @aljoscha, I have switched to other ongoing things. I'll discuss with Aviem about that particular point.

**Comment by echauchot** [comment]
Run Apex ValidatesRunner

**Comment by echauchot** [comment]
Run Gearpump ValidatesRunner

**Comment by echauchot** [comment]
Run Flink ValidatesRunner

**Comment by echauchot** [comment]
Seems that after all the coder is not the problem: the problem is that Spark runner casts the CustomWindow to IntervalWindow in the merge process, hence the serialization issue with the CustomWindow downstream. I'm investigating how to fix that up.

**Comment by jbonofre** [comment]
Gonna take a look.

R: @jbonofre

**Comment by echauchot** [comment]
I opened an issue on the spark runner side to fix custom window merging (https://issues.apache.org/jira/browse/BEAM-2499)
R: @amitsela I add you as a reviewer on this PR so that you can give your opinion on whether this `ValidatesRunner` test is valid.
@aljoscha, @amitsela, maybe we could merge this PR and use this new test to ensure that all the runners support Custom Window merging?

**Comment by etienne-chauchot** [comment]
@aljoscha @amitsela WDYT?

**Comment by echauchot** [comment]
@aljoscha @amitsela WDYT?

**Comment by aljoscha** [comment]
From my side the test looks good. We could think about adding it to `WindowTest`, instead of creating a new file, however.

Regarding the Spark runner, this needs either an exclusion or a fix.

**Comment by echauchot** [comment]
Thanks @aljoscha. I agree, it is better to put this test with the other window related tests. I'll also add a new category interface to skip this test in runners that don't support it

**Comment by echauchot** [comment]
Done and rebased on master

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/UsesCustomWindowMerging.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.testing;
+
+import org.apache.beam.sdk.transforms.ParDo;
+
+/**
+ * Category tag for validation tests which utilize splittable {@link ParDo}.
```
Comment (aljoscha) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/UsesCustomWindowMerging.java]: I think this comment was not properly adapted.

**Comment by aljoscha** [comment]
Run Flink ValidatesRunner

**Comment by aljoscha** [comment]
Run Spark ValidatesRunner

**Comment by aljoscha** [comment]
retest this please

**Comment by aljoscha** [comment]
Run Flink ValidatesRunner

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/UsesCustomWindowMerging.java (hunk 1 above)
Comment (echauchot) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/UsesCustomWindowMerging.java]: My bad, bad copy/paste :) Thanks for pointing !

**Comment by echauchot** [comment]
Spark Runner Validates runner failed with `connection reset on https://repo.maven.apache.org:443`

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by echauchot** [comment]
This PR is to be merged before https://github.com/apache/beam/issues/3517 because it adds the that is used in  https://github.com/apache/beam/issues/3517

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by echauchot** [comment]
Spark Runner ValidatesRunner tests errors above are not related to this PR, they are not related to custom window merging. I think we could merge the PR if you are ok. 
WDYT?

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by echauchot** [comment]
Run Flink ValidatesRunner

**Comment by echauchot** [comment]
Run Flink ValidatesRunner

**Comment by amitsela** [comment]
LGTM on excluding `SparkRunner`

**Comment by echauchot** [comment]
Above validatesRunner tests failures are not related to this PR. See these tickets:
https://issues.apache.org/jira/browse/BEAM-2571
https://issues.apache.org/jira/browse/BEAM-2670
https://issues.apache.org/jira/browse/BEAM-2671

**Comment by echauchot** [comment]
R: @aviemzur

**Comment by echauchot** [comment]
Run Flink ValidatesRunner

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by aviemzur** [comment]
Merged based on @amitsela LGTM

**Pull request closed without merging**